src/backend/distributed/planner/multi_physical_planner.c

```diff
@@ -1699,41 +1701,55 @@ ChildNodeList(MultiNode *multiNode)
 
 /*
  * UniqueJobId allocates and returns a unique jobId for the job to be executed.
- * This allocation occurs both in shared memory and in write ahead logs; writing
- * to logs avoids the risk of having jobId collisions.
- *
- * Please note that the jobId sequence wraps around after 2^32 integers. This
- * leaves the upper 32-bits to slave nodes and their jobs.
  */
 static uint64
 UniqueJobId(void)
 {
-	text *sequenceName = cstring_to_text(JOBID_SEQUENCE_NAME);
-	Oid sequenceId = ResolveRelationId(sequenceName);
-	Datum sequenceIdDatum = ObjectIdGetDatum(sequenceId);
-	Datum jobIdDatum = 0;
-	int64 jobId = 0;
-	int64 localizedJobId = 0;
-	int64 localGroupId = GetLocalGroupId();
-	Oid savedUserId = InvalidOid;
-	int savedSecurityContext = 0;
-
-	GetUserIdAndSecContext(&savedUserId, &savedSecurityContext);
-	SetUserIdAndSecContext(CitusExtensionOwner(), SECURITY_LOCAL_USERID_CHANGE);
-
-	/* generate new and unique jobId from sequence */
-	jobIdDatum = DirectFunctionCall1(nextval_oid, sequenceIdDatum);
-	jobId = DatumGetInt64(jobIdDatum);
+	static uint32 jobIdCounter = 0;
+
+	uint64 jobId = 0;
+	uint64 jobIdNumber = 0;
+	uint64 processId = 0;
+	uint64 localGroupId = 0;
+
+	jobIdCounter++;
+
+	if (EnableUniqueJobIds)
+	{
+		/*
+		 * Add the local group id information to the jobId to
+		 * prevent concurrent jobs on different groups to conflict.
+		 */
+		localGroupId = GetLocalGroupId() & 0xFF;
+		jobId = jobId | (localGroupId << 48);
+
+		/*
+		 * Add the current process ID to distinguish jobs by this
+		 * backends from jobs started by other backends. Process
+		 * IDs can have at most 24-bits on platforms supported by
+		 * Citus.
+		 */
+		processId = MyProcPid & 0xFFFFFF;
+		jobId = jobId | (processId << 24);
```

Comment:
Perhaps it'd be better to use BackendId instead of MyProcPid?  Every backend has one, and they only get re-assigned after a backend has shutdown. Has the advantage of being smaller...